lib/api/core/models/repositories/repository.js

```diff
@@ -1,344 +1,421 @@
+   *
+   * @param {Profile|Role|User} object - The object to persist
+   * @param {object} [options] - The persistence options
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  persistToDatabase (object, options = {}, collection = this.collection) {
+    const method = options.method || 'createOrReplace';
+
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
+    }
+
+    return this.databaseEngine[method](collection, object._id, this.serializeToDatabase(object));
   }
 
-  return Promise.all(promises);
-};
+  /**
+   * Given an object id, delete it from the configured storage engines
+   *
+   * @param {string} id - The id of the object to delete
+   * The options optional parameters currently accepts only 1 option:
+   *   key: if provided, removes the given key instead of the default one (<collection>/<id>)
+   * @param {object} [options] - optional options for the current operation
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  delete (id, options = {}, collection = this.collection) {
+    const promises = [];
+
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
+    }
 
+    if (this.cacheEngine) {
+      promises.push(this.deleteFromCache(id, options, collection));
+    }
 
-/**
- * Delete repository from database according to its id
- * @param id
- */
-Repository.prototype.deleteFromDatabase = function repositoryDeleteFromDatabase (id) {
-  return this.databaseEngine.delete(this.collection, id);
-};
+    if (this.databaseEngine) {
+      promises.push(this.deleteFromDatabase(id, collection));
+    }
 
-/**
- * Persists the given ObjectConstructor object in cache.
- * The opts optional parameters currently accept 2 options:
- *   key: if provided, stores the object to the given key instead of the default one (<collection>/<id>)
- *   ttl: if provided, overrides the default ttl set on the repository for the current operation.
- *
- * @param {object} object the object to persist
- * @param {object} [opts] optional options for the current operation
- * @returns {Promise}
- */
-Repository.prototype.persistToCache = function repositoryPersistToCache (object, opts) {
-  var
-    options = opts || {},
-    key = options.key || this.cacheKeyPrefix + object._id,
-    ttl = this.ttl;
-
-  if (options.ttl !== undefined) {
-    ttl = options.ttl;
+    return Promise.all(promises);
   }
-  if (ttl === false) {
-    return this.cacheEngine.set(key, JSON.stringify(this.serializeToCache(object)))
+
+
+  /**
+   * Delete repository from database according to its id
+   *
+   * @param id
+   * @param {string} [collection]
+   */
+  deleteFromDatabase (id, collection = this.collection) {
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
```

Comment:
It can not happen due to a bad request, it might come from an bad plugin implementation, will use this one instead.